core/src/apps/cardano/get_address.py

```diff
@@ -1,31 +1,108 @@
+    ctx: wire.Context, msg: CardanoGetAddress, keychain: seed.Keychain
+) -> CardanoAddress:
+    address_parameters = msg.address_parameters
+
+    await paths.validate_path(
+        ctx, validate_full_path, keychain, address_parameters.spending_key_path, CURVE
+    )
 
     try:
-        address, _ = derive_address_and_node(
-            keychain, msg.address_n, msg.protocol_magic
+        address = derive_human_readable_address(
+            keychain, address_parameters, msg.protocol_magic, msg.network_id
         )
     except ValueError as e:
         if __debug__:
             log.exception(__name__, e)
         raise wire.ProcessError("Deriving address failed")
+
     if msg.show_display:
-        desc = address_n_to_str(msg.address_n)
-        while True:
-            if await show_address(ctx, address, desc=desc):
-                break
-            if await show_qr(ctx, address, desc=desc):
-                break
+        await _display_address(
+            ctx, keychain, address_parameters, address, msg.protocol_magic
+        )
 
     return CardanoAddress(address=address)
+
+
+async def _display_address(
+    ctx: wire.Context,
+    keychain: seed.Keychain,
+    address_parameters: CardanoAddressParametersType,
+    address: str,
+    protocol_magic: int,
+) -> None:
+    await _show_staking_warnings(ctx, keychain, address_parameters)
+
+    network = None
+    if not protocol_magics.is_mainnet(protocol_magic):
+        network = protocol_magic
+
+    while True:
+        if await show_address(
+            ctx,
+            address,
+            address_parameters.address_type,
+            address_parameters.spending_key_path,
+            network=network,
+        ):
+            break
+        if await show_qr(
+            ctx, address, desc=address_n_to_str(address_parameters.spending_key_path)
+        ):
+            break
+
+
+async def _show_staking_warnings(
+    ctx: wire.Context,
+    keychain: seed.Keychain,
+    address_parameters: CardanoAddressParametersType,
+) -> None:
+    staking_type = staking_use_cases.get(keychain, address_parameters)
+    if staking_type == staking_use_cases.DIFFERENT_ACCOUNT:
+        await show_warning_address_foreign_staking_key(
+            ctx,
+            _to_account_path(address_parameters.spending_key_path),
+            _to_account_path(address_parameters.staking_key_path),
+            None,
+        )
+    elif staking_type == staking_use_cases.DIFFERENT_HASH:
+        await show_warning_address_foreign_staking_key(
+            ctx,
+            _to_account_path(address_parameters.spending_key_path),
+            None,
+            address_parameters.staking_key_hash,
+        )
+    elif staking_type == staking_use_cases.POINTER_ADDRESS:
+        await show_warning_address_pointer(ctx, address_parameters.certificate_pointer)
+
+
+def _to_account_path(path: List[int]) -> List[int]:
+    ACCOUNT_PATH_LENGTH = const(3)
+
+    if len(path) < ACCOUNT_PATH_LENGTH:
```

Comment:
can this happen?

If yes, this is not the place to fail for the problem.

either:
a) make the `validate_full_path` above a mandatory check (instead of `validate_path`, just call `if not validate_full_path(...): raise wire.DataError`)
b) make this code resilient to the shortness - perhaps just removing this check is OK?
c) make `staking_use_cases` recognize a new state like "paths could not be analyzed"? and display the appropriate warning.

FWIW making the path schema mandatory is a perfectly OK thing to do, unless you have a usecase for custom paths in mind.

Reply:
I've removed the check [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/3831a7ab4d0707a00760c25ba3696494a26fe541). All the paths are validated and even if not, this wouldn't cause any critical error.